### ISSUE TITLE
Does Fastify v3 support supertest?

### ISSUE THREAD

**Issue by saulotoledo** [fastify/help#228]
I recently updated my project to support Fastify v3, but all my tests are broken.
I created a repository to replicate the issue: https://github.com/saulotoledo/nestjs-fastify-testing-issue

#### What was the result you received?

In that project, I have a NestJS application. It is possible to see the issue by running `yarn test:e2e`. The test will fail because I am using the Fastify adapter. The following error is triggered by fastify:

```bash
TypeError: Cannot read property 'length' of undefined
    at next (/Users/saulos/sample-nest-app/node_modules/fastify/lib/route.js:374:32)
    at preParsingHookRunner (/Users/saulos/sample-nest-app/node_modules/fastify/lib/route.js:405:3)
    at runPreParsing (/Users/saulos/sample-nest-app/node_modules/fastify/lib/route.js:356:5)
    at Object.routeHandler [as handler] (/Users/saulos/sample-nest-app/node_modules/fastify/lib/route.js:334:7)
    at Router.lookup (/Users/saulos/sample-nest-app/node_modules/fastify/node_modules/find-my-way/index.js:351:14)
    at Server.emit (events.js:203:13)
    at parserOnIncoming (_http_server.js:713:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:116:17)
```
#### What did you expect?

The test should run.

#### Context

* *node version*: 12
* *fastify version*: >=3.1.1
* *os*: Mac

**Comment by mcollina** [comment]
I'm not a user of supertest nor we support it directly :/, but it should work. I recommend to use our inject utility instead: https://www.fastify.io/docs/latest/Testing/#benefits-of-using-fastifyinject.

**Comment by mcollina** [comment]
Happy to review a PR that fixes the supertest support, if broken.

**Comment by saulotoledo** [comment]
@mcollina I just dropped supertest for now and started using fastify inject while I can. I believe supertest support is broken, but I am not sure where is the problem. This week is being overwhelming, I am not sure if I will be able to investigate that this time, but I will do it if I can.

I am having another issue with inject: I am not able to test [fastify-helmet](https://github.com/fastify/fastify-helmet). Helmet must be used before any other `app.use()`. Does inject add something, so helmet does not work?

**Comment by Eomm** [comment]
> app.use()

Could you show an example?
I think you should not use `app.use` with plugins, but `app.register` in order to let them work

> so helmet does not work?

No, it should work as well, we used inject in the test too https://github.com/fastify/fastify-helmet/blob/master/test.js

**Comment by RafaelGSS** [comment]
TL;DR:
Change your `await app.init()` to: `await new Promise(resolve => app.listen(3535, resolve));`

I don't know how `init` is wrapped at nest package, but your tests (with `app.init()`) are the following:

- Register your route [here](https://github.com/fastify/fastify/blob/v3.1.1/lib/route.js#L216)
- Receive a request and on [this statement](https://github.com/fastify/fastify/blob/v3.1.1/lib/reply.js#L322) are considering `true` because `preParsing` are undefined not null

`preParsing` is undefined because [this statement](https://github.com/fastify/fastify/blob/v3.1.1/lib/route.js#L243) not run. And this is your unexpected behavior. Maybe you can check the `init` at `nestjs` to fix for them :smile:

**Comment by saulotoledo** [comment]
@Eomm I was trying to isolate the problem in an example, and the first time I got it was with some broken promises in this repository: https://github.com/saulotoledo/nestjs-fastify-helmet-not-working

I did not understand (I might be missing something) what is the relationship between that promise and the plugin working or not. If you fix the promise, you can see the proper headers. To run the e2e test in that repository: `yarn test:e2e`.

@RafaelGSS Thanks for the information! It will be helpful. I will see what I can do. :D

**Comment by RafaelGSS** [comment]
@saulotoledo I don't get the issue.

I run here and the tests work fine. Can you add another test case to explain?

**Comment by saulotoledo** [comment]
@RafaelGSS I improvised a bit, but the `console.log()` output is something like this, right?

```json
{
      connection: 'close',
      'content-type': 'application/json',
      'content-length': '80',
      date: 'Wed, 05 Aug 2020 02:17:15 GMT'
}
```

If yes, this is not correct. We should have some headers from helmet in addition to those:

```json
{
      'x-dns-prefetch-control': 'off',
      'x-frame-options': 'SAMEORIGIN',
      'strict-transport-security': 'max-age=15552000; includeSubDomains',
      'x-download-options': 'noopen',
      'x-content-type-options': 'nosniff',
      'x-xss-protection': '1; mode=block',
      connection: 'close',
      'content-type': 'application/json',
      'content-length': '80',
      date: 'Wed, 05 Aug 2020 02:17:15 GMT'
}
```

I will update the repository asap, but could you confirm if you see what I've described here?

**Comment by RafaelGSS** [comment]
change the:

```diff
- await app.init();
- server = await app.listen(3003);
+ await new Promise(resolve => app.listen(3003, resolve));
```

is not the same problem as described on the issue, but the behaviour is the same.

**Comment by saulotoledo** [comment]
@RafaelGSS Thanks for double checking that. I was trying to solve the problem as in the docs, but now I am more convinced that I need some workaround. For now this is working, I just need to solve some open handles in Jest and I am ready to the next step.

**Comment by joticajulian** [comment]
We had the same issue and we solved it using 
```
await app.init();
await app.getHttpAdapter().getInstance().ready();
```

**Comment by Abszissex** [comment]
@joticajulian 
thanks a lot, this one made my tests work again. Real lifesaver :-)

**Comment by lvkins** [comment]
In my case adding a `await app.listen(...)` after `await app.init()` resolved mentioned issue.

```ts
await app.init();
await app.listen(3000, '0.0.0.0');
```
